packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -49,50 +57,78 @@ class ProtocolDiscovery {
   final int devicePort;
   final bool ipv6;
 
-  final Completer<Uri> _completer = Completer<Uri>();
+  /// The time to wait before forwarding observatory URIs from [logReader].
+  final int throttleTimeInMilliseconds;
 
   StreamSubscription<String> _deviceLogSubscription;
 
+  StreamController<Uri> _uriStreamController;
+
   /// The discovered service URI.
+  /// Use [uris] instead.
+  // TODO(egarciad): replace `uri` for `uris`.
+  Future<Uri> get uri async {
+    try {
+      return await uris.first;
+    } on StateError {
```

Comment:
This seems like a bug. We generally shouldn't be catching Error subclasses unless it is a temporary work around. If you are going to replace uri with uris I would do that in this PR

Reply:
I'm not sure I follow. If something called cancel too early, then this is the equivalent behavior in the current code.

Reply:
In what situation does `_uriStreamController?.close()` triggers the `StateError`? When you say `uris.first` after the `_uriStreamController?.close()`? Can you check `_uriStreamController.isClosed` or add an `onCancel` callback to `_uriStreamController` that records a bit?

Reply:
The situation is that there's a test **returns non-null uri future** under `protocol_discovery_test.dart` that expects `uri` to return a non-null value. For some reason, this works currently even though the `uri` future never completes. I checked the documentation for `expect` and it says:
" If  [actual] is a [Future], the test won't complete until the future emits a value."  However, it doesn't seem true in this case:
```.dart
final Completer<void> _completer = Completer<Uri>();
expect(_completer.future, isNotNull);
```


For this reason, the test is failing when migrated to streams.  The stream is closed in the `tearDown` callback before a value is emitted.

Reply:
catching a StateError is definitely a bug. It implies there's some logic error in our code.